arrows/ffmpeg/ffmpeg_video_input.cxx

```diff
@@ -1001,14 +1001,24 @@ ::frame_image( )
     bool direct_copy;
     AVFrame* frame = d->f_frame;
 
+    class frame_ref
```

Comment:
Is this class necessary? I feel like it adds unneeded complexity where the same thing could be achieved by just adding the `av_frame_unref(d->filter_frame)` call at the appropriate place.

Reply:
That was my thought exactly.  If the missing `av_frame_unref` was the culprit it would seem that this fix could be achieved by just adding that one line.

Reply:
Yes, RAII is necessary. There are something like 4-5 places that need to unref, and getting them right is tricky (I think this is why my first attempt at just using unref didn't work). RAII is ***way*** less error-prone.

That said, if you know of a usable, generic RAII class, I'm all ears. Note however we have the quirky trait of needing to trigger it early *without* "resetting" it (i.e. it still needs to fire on destruction).

Reply:
Also, it's actually *more* lines of code to not use this helper, and that's leaving aside that not using RAII here is far more brittle.

```diff
diff --git a/arrows/ffmpeg/ffmpeg_video_input.cxx b/arrows/ffmpeg/ffmpeg_video_input.cxx
index 30f8e4e8c..c14b94e27 100644
--- a/arrows/ffmpeg/ffmpeg_video_input.cxx
+++ b/arrows/ffmpeg/ffmpeg_video_input.cxx
@@ -1000,18 +1000,7 @@ ::frame_image( )
     vital::image_pixel_traits pixel_trait = vital::image_pixel_traits_of<unsigned char>();
     bool direct_copy;
     AVFrame* frame = d->f_frame;
-
-    class frame_ref
-    {
-    public:
-      ~frame_ref() { unref(); }
-      void operator=(AVFrame* frame) { unref(); m_frame = frame; }
-
-    private:
-      void unref() { if (m_frame) av_frame_unref(m_frame); }
-
-      AVFrame* m_frame = nullptr;
-    } filtered_frame_ref;
+    AVFrame* filtered_frame = nullptr;
 
     if (d->f_filter_src_context && d->f_filter_sink_context)
     {
@@ -1025,14 +1014,23 @@ ::frame_image( )
           AV_BUFFERSRC_FLAG_KEEP_REF) < 0)
         {
           LOG_ERROR(this->logger(), "Error while feeding the filter graph");
+          if (filtered_frame)
+          {
+            av_frame_unref(filtered_frame);
+          }
           return nullptr;
         }
         // Pull a filtered frame from the filter graph
-        filtered_frame_ref = d->f_filtered_frame;
+        if (filtered_frame)
+        {
+          av_frame_unref(filtered_frame);
+        }
+        filtered_frame = d->f_filtered_frame;
         auto const ret = av_buffersink_get_frame(d->f_filter_sink_context,
                                                  d->f_filtered_frame);
         if (ret == AVERROR_EOF)
         {
+          av_frame_unref(filtered_frame);
           return nullptr;
         }
         if (ret == AVERROR(EAGAIN))
@@ -1109,6 +1107,10 @@ ::frame_image( )
       if (!d->f_software_context)
       {
         LOG_ERROR(this->logger(), "Couldn't create conversion context");
+        if (filtered_frame)
+        {
+          av_frame_unref(filtered_frame);
+        }
         return nullptr;
       }
 
@@ -1130,6 +1132,10 @@ ::frame_image( )
       depth, depth * width, 1
     );
     d->current_image = std::make_shared<vital::simple_image_container>(vital::simple_image_container(image));
+    if (filtered_frame)
+    {
+      av_frame_unref(filtered_frame);
+    }
   }
 
   return d->current_image;
```

Reply:
I'm not sure what you are trying to say here.  This is no longer about RAII.  Sure, we could write a bunch of wrappers around FFMPEG classes to make them easier to use safely, but that's not the point of this branch.  We just need a one-line fix here.

In this case, I think `unref` is poorly named and really means "reset the internal state of this class".  I wouldn't have thought we needed to do that explicitly, here, but apparently we do.

It's actually somewhat risky adding this extra class, because you are saying that every time we implicitly convert an `AVFrame*` into this wrapper class it is going to reset the internal state of the class, which is not very intuitive and would be easy to accidently clear the data you still need.

Reply:
okay, I see the confusion.  We shouldn't need to (and probably don't want to) `unref` in those places where your `filtered_frame` goes out of scope.  In fact, this could destroy the data we successfully read by doing this.  The `unref` will eventually happen when the next frame is read.  The memory is not really leaking, it just seems that the AVFrame likes to keep old data around until we manually tell it to clear its buffers, which we were previously not doing at all.

By the way, what version of FFMPEG do you have?  Maybe this is an FFMPEG bug or change in how FFMPEG works.  I'm surprised that I haven't see this data hording behavior before.

Reply:
> This is no longer about RAII.

The problem is we need to free a resource and *we don't know when it goes out of scope*. That is just about the *definition* of why RAII exists.

> We just need a one-line fix here.

Okay, okay... it looks like it's possible to twist `std::unique_ptr` into doing the same thing. (I originally didn't think that was going to be possible without at least as much additional code, and possibly memory allocations also.) All that's doing though is using a standard type instead of the custom type; the behavior of the two is the same. (Although, not needing a custom type is good.)

> every time we implicitly convert an `AVFrame*` [...] [it] would be easy to accidentally clear the data you still need.

This exact same argument could be applied to `std::unique_ptr`, but no one makes that argument because obviously you don't assign to a `unique_ptr` unless you're done with whatever it had previously.

`frame_ref` has the usage semantics of `std::unique_ptr`. And now that it is replaced with actually `std::unique_ptr`, hopefully that is more clear.

Reply:
In version without the RAII wrapper, why do you introduce an additional AVFrame pointer (`filtered_frame`) when you can just use `d->f_filtered_frame` directly? Currently `d->f_filtered_frame` is allocated in `open`, I wonder if it makes since to allocate it and `f_frame` in the constructor. Then the `if (filtered_frame)` guards may not be needed. I know that the ffmpeg `_free` methods all check before they try and free memory so they don't need guard blocks. Looking at the source code for `av_frame_unref` that does not seem to be the case for that method.

Reply:
Is it safe to unref something that has never been "allocated"? Otherwise we would need to add an additional member whether the filtered frame has ever been used, and possibly an `av_frame_unref` prior to the `av_frame_free` (it's unclear if this is necessary, but at least one source I was looking at seems to think it is). Otherwise we're still leaking memory, if only one frame per `ffmpeg_video_input` constructed instead of (at least?) one frame for every frame read.

I don't see how data loss could occur. AFAICT, we only ever look at `f_filtered_frame` if we have first called `av_buffersink_get_frame`, and we either successfully copy it into a `vital::image_memory`, or we don't and the next time we might look at it, we will have again called `av_buffersink_get_frame` first. I'd rather free it sooner than hang on to it for no significant benefit. (This is _not_ the case for `f_frame`, but that's not the one that's leaking.)

I'm using ffmpeg-4.3.1-16.fc33.x86_64.

Reply:
> why do you introduce an additional AVFrame pointer (`filtered_frame`) when you can just use `d->f_filtered_frame` directly?

...Because I can't? That isn't set if `open` has been called, it is set if `av_buffersink_get_frame` was called, which in particular may not happen because that entire block might not execute. I could have equally written `bool av_buffersink_get_frame_was_called = false` and then, yes, used `av_frame_unref(d->f_filtered_frame)`.

Reply:
After a more careful read of the code I now see that your approach should work without any side effects.  I was thinking that this filtered frame was reused later in a different function, but that's not the case.

My earlier point was that we could make a one line fix by calling `av_frame_unref` right before `av_buffer_sink_get_frame` to free up any memory from a previous call to this function.  I think either approach would work, but what you are proposing does the cleanup sooner, so that's probably better (as long as we don't add any functions that expect the last filtered frame to still be valid outside of this function call).

The `f_frame` does not leak because `avcodec_receive_frame` interally calls `av_frame_unref` where as `av_buffer_sink_get_frame` does not.  That behavior is documented, though I don't know why it differs between functions that get frames.  In the case of `f_frame` we do share that result between function calls so we would have to be careful not to unref that one too early.